R/attr.R

```diff
@@ -0,0 +1,174 @@
+#' \code{""}). \code{is_dict()} checks that an object is a dictionary:
+#' that it has actual names and in addition that there are no
+#' duplicated names. \code{has_names()} is a vectorised version of
+#' \code{is_named()}.
+#'
+#' @param x An object to test.
+#' @return \code{is_named()} and \code{is_dict()} are scalar
+#'   predicates and return \code{TRUE} or \code{FALSE}.
+#'   \code{has_names()} is vectorised and returns a logical vector as
+#'   long as the input.
+#' @export
+#' @examples
+#' # A data frame usually has valid, unique names
+#' is_named(mtcars)
+#' has_names(mtcars)
+#' is_dict(mtcars)
+#'
+#' # But data frames can also have duplicated columns:
+#' dups <- cbind(mtcars, cyl = seq_len(nrow(mtcars)))
+#' is_dict(dups)
+#'
+#' # The names are still valid:
+#' is_named(dups)
+#' has_names(dups)
+#'
+#'
+#' # For empty objects the semantics are slightly different.
+#' # is_dict() returns TRUE for empty objects:
+#' is_dict(list())
+#'
+#' # But is_named() will only return TRUE if there is a names
+#' # attribute (a zero-length character vector in this case):
+#' x <- set_names(list(), character(0))
+#' is_named(x)
+#'
+#'
+#' # Empty and missing names are invalid:
+#' invalid <- dups
+#' names(invalid)[2] <- ""
+#' names(invalid)[5] <- NA
+#'
+#' # is_named() performs a global check while has_names() can show you
+#' # where the problem is:
+#' is_named(invalid)
+#' has_names(invalid)
+#'
+#' # has_names() will work even with vectors that don't have a names
+#' # attribute:
+#' has_names(letters)
+is_named <- function(x) {
+  nms <- names(x)
+
+  if (is_null(nms)) {
+    return(FALSE)
+  }
+
+  if (any(nms == "" | is.na(nms))) {
+    return(FALSE)
+  }
+
+  TRUE
+}
+
+#' @rdname is_named
+#' @export
+is_dict <- function(x) {
+  if (!length(x)) {
+    return(!is.null(x))
+  }
+
+  is_named(x) && !any(duplicated(names(x)))
+}
+
+#' @rdname is_named
+#' @export
+has_names <- function(x) {
+  nms <- names(x)
+  if (is.null(nms)) {
+    rep(FALSE, length(x))
+  } else {
+    !(is.na(nms) | nms == "")
+  }
+}
+
+#' Set names of a vector.
+#'
+#' This is a snake case wrapper for \code{\link[stats]{setNames}},
+#' with tweaked defaults, and stricter argument checking.
+#'
+#' @param x Vector to name.
+#' @param nm Vector of names, the same length as \code{x}.
+#' @export
+#' @examples
+#' set_names(1:4, c("a", "b", "c", "d"))
+#'
+#' # If the second argument is ommitted a vector is named with itself
+#' set_names(letters[1:5])
+set_names <- function(x, nm = x) {
+  if (!is_vector(x)) {
+    stop("`x` must be a vector", call. = FALSE)
+  }
+  if (!is_null(nm) && length(x) != length(nm)) {
+    stop("`x` and `nm` must be the same length", call. = FALSE)
+  }
+
+  names(x) <- nm
+  x
+}
+
+#' Get names of a vector.
+#'
+#' This names getter always returns a character vector, even when an
+#' object does not have a \code{names} attribute. In this case, it
+#' returns a vector of empty names \code{""}.
```

Comment:
since `names2()` is usually not implemented that way, it should probably be an option. `regularise_na = FALSE`?